File: xylitol.py
#!/usr/bin/python
# -*- encoding: utf-8 -*-

import socket, re, hashlib, sys

class AddrSpec:

  def __init__(self, buf):
    m = re.match(r'([^:]+):', buf)
    self.scheme, buf = m.group(1), buf[m.end():]
    if self.scheme == 'sip':
      m = re.match(r'(([^@]+)@)?([^?;]+)', buf)
      self.userinfo, hostport = m.group(2), m.group(3)
      buf = buf[m.end():]
      m = re.match(r'([^:]+)(:(\d+))?', hostport)
      self.host, self.port = m.group(1), m.group(3)
    else:
      self.host = self.port = None
    m = re.match(r'[^?]*', buf)
    self.uri_prms, self.headers = m.group(0), buf[m.end():]


class NameAddr(AddrSpec):

  def __init__(self, buf):
    m = re.match(r'(("[^"]*")\s*)?', buf)
    self.display_name, buf = m.group(0), buf[m.end():]
    m = re.match(r'\s*<([^>]+)>', buf)
    if m != None:
      self.addr_spec, buf = m.group(1), buf[m.end():]
    else:
      m = re.match(r'[^;]+', buf)
      self.addr_spec, buf = m.group(0), buf[m.end():]
    self.prms = buf
    AddrSpec.__init__(self, self.addr_spec)


class Header:

  def __init__(self, name, vals):
    self.name, self.vals = name, vals

  def __str__(self):
    return self.name + ': ' + ', '.join(self.vals) + "\r\n"


class Message:

  def __init__(self, buf):
    buf = re.sub(r'^((\r\n)|(\r)|(\n))*', "", buf)
    m = re.search(r'((\r\n\r\n)|(\r\r)|(\n\n))', buf)
    self.body = buf[m.end():]
    buf = re.sub(r'\n[ \t]+',' ', re.sub(r'\r\n?', "\n", buf[:m.start()]))
    ary = buf.split("\n")
    m = re.match(r'(([A-Z]+) ([^ ]+) )?SIP\/2\.0( (\d+) ([^\n]+))?', ary[0])
    self.method, self.requri, self.stcode, self.reason = \
      m.group(2), m.group(3), m.group(5), m.group(6)
    self.hdrs = []
    for buf in ary[1:]:
      name, buf = re.split(r'\s*:\s*', buf, 1)
      self.hdrs.append(Header(name, re.split(r'\s*,\s*', buf)))

  def __str__(self):
    if self.method != None:
      s = self.method + ' ' + self.requri + " SIP/2.0\r\n"
    else:
      s = 'SIP/2.0 ' + self.stcode + " " + self.reason + "\r\n"
    for hdr in self.hdrs:
      s += str(hdr)
    return s + "\r\n" + self.body

  def search(self, name1, name2 = ''):
    for i, h in enumerate(self.hdrs):
      if name1.lower() == h.name.lower() or name2.lower() == h.name.lower():
        return i
    return None

  def rsearch(self, name1, name2 = ''):
    pos = None
    for i, h in enumerate(self.hdrs):
      if name1.lower() == h.name.lower() or name2.lower() == h.name.lower():
        pos = i
    return pos

  def gen_resp(self, stcode, reason, contacts = []):
    hs=["call-id","i","from","f","to","t","via","v","cseq","record-route"]
    resp = Message("SIP/2.0 " + stcode + " " + reason + "\r\n\r\n")
    for h in self.hdrs:
      if h.name.lower() in hs:
        resp.hdrs.append(h)
    if contacts != []:
      resp.hdrs.append(Header("Contact", contacts))
    resp.hdrs.append(Header("Content-Length", ["0"]))
    return resp


class Proxy:

  def __init__(self, domain, ip, port):
    self.location_service = {}
    self.domain, self.ip, self.port = domain, ip, port
    self.via = "SIP/2.0/UDP " + self.ip + ":" + str(self.port) + ";branch="
    self.rr = "<sip:" + self.ip + ":" + str(self.port) + ";lr>"
    self.sr = "<sip:" + self.ip + ":" + str(self.port) + ">"
    self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    self.sock.bind((self.ip, self.port))

  def start(self):
    while True:
      buf, addr = self.sock.recvfrom(0xffff)
      print "<" * 72
      print "received from " + addr[0] + " " + str(addr[1])
      print "-" * 8
      print buf
      msg = Message(buf)
      if msg.method != None:
        viapos = msg.search("via", "v")
        msg.hdrs[viapos].vals[0] += ";received=" + addr[0]
        m = re.search(";\s*rport", msg.hdrs[viapos].vals[0])
        if m != None:
          tmpvia = msg.hdrs[viapos].vals[0]
          tmpvia = tmpvia[:m.end()] + "=" + str(addr[1]) + tmpvia[m.end():]
          msg.hdrs[viapos].vals[0] = tmpvia
        branch = msg.hdrs[viapos].vals[0] + " "
        branch += msg.hdrs[msg.search("call-id", "i")].vals[0] + " "
        cseq = msg.hdrs[msg.search("cseq", "")].vals[0]
        m = re.match(r'(\d+)\s+(\S+)', msg.hdrs[msg.search("cseq")].vals[0])
        cseq_num, cseq_method = m.group(1), m.group(2)
        if cseq_method == "ACK":
          cseq_method = "INVITE"
        branch += cseq_num + " " + cseq_method
        branch = 'z9hG4bK' + hashlib.md5(branch).hexdigest()
        msg.hdrs[viapos].vals.insert(0, self.via + branch)
        self.proc_request(msg)
      else:
        self.proc_response(msg)

  def proc_request(self, msg):
    # (1) Proxy-Requireがあったらエラー
    pos = msg.search("proxy-require")
    if pos != None:
      if msg.method != "ACK":
        unsupported = msg.hdrs[pos]
        resp = msg.gen_resp("420", "Bad Extension")
        resp.hdrs.append(unsupported)
        proc_response(resp)
      return
    # (2) Request-URIが自身を指しているかを判定する
    requri = AddrSpec(msg.requri)
    if self.comp(requri, self.domain, self.ip, self.port):
      if msg.method == "REGISTER":
        return self.proc_register(msg)
      if self.location_service.has_key(requri.userinfo):
        msg.requri = self.location_service[requri.userinfo]
      else:
        if msg.method != "ACK":
          self.proc_response(msg.gen_resp("404", "Not Found"))
        return
      requri = AddrSpec(msg.requri)
    # (3) Max-Forwardsの確認
    pos = msg.search("max-forwards")
    if pos == None:
      msg.hdrs.append(Header("Max-Forwards", ["70"]))
    elif msg.hdrs[pos].vals[0] == "0":
      if msg.method != "ACK":
        resp = msg.gen_resp("420", "Too Many Hops")
        self.proc_response(resp)
      return
    else:
      msg.hdrs[pos].vals[0] = str(int(msg.hdrs[pos].vals[0]) -1)
    # (4) Record-Routeヘッダを処理
    pos = msg.rsearch("record-route")
    if pos == None:
      msg.hdrs.append(Header("Record-Route", [self.rr]))
    else:
      msg.hdrs[pos].vals.append(self.rr)
    # (5) Routeヘッダの先頭が自分だったら削除する
    pos = msg.search("route")
    if pos != None:
      route = NameAddr(msg.hdrs[pos].vals[0])
      if self.comp(route, self.domain, self.ip, self.port):
        del msg.hdrs[pos].vals[0]
        if msg.hdrs[pos].vals == []:
          del msg.hdrs[pos]
    # (6) 送信先判定、Routeヘッダ値、なければRequest-URIで判定
    pos = msg.search("route")
    if pos != None:
      target = NameAddr(msg.hdrs[pos].vals[0])
    else:
      target = requri
    self.send(str(msg), target.host, target.port)

  def proc_register(self, msg):
    addr = NameAddr(msg.hdrs[msg.search("to", "t")].vals[0])
    contact = msg.hdrs[msg.search("contact", "m")].vals[0]
    self.location_service[addr.userinfo] = contact
    resp = msg.gen_resp("200", "OK", [contact])
    resp.hdrs.append(Header("Service-Route", [self.sr]))
    self.proc_response(resp)

  def proc_response(self, msg):
    # (1) 先頭のViaを削除
    pos = msg.search("via", "v")
    del msg.hdrs[pos].vals[0]
    if msg.hdrs[pos].vals == []:
      del msg.hdrs[pos]
      pos = msg.search("via", "v")
    # (2) 中継先判定
    m = re.match(r'SIP\s*\/\s*2\.0\s*\/\s*UDP\s+([^\s;:]+)(\s*:\s*(\d+))?',
      msg.hdrs[pos].vals[0])
    desthost, destport = m.group(1), m.group(3)
    prms = msg.hdrs[pos].vals[0][m.end():]
    m = re.search(r';\s*received\s*=\s*([^\s;]+)', prms)
    if m != None:
      desthost = m.group(1)
    m = re.search(r';\s*rport\s*=\s*(\d+)', prms)
    if m != None:
      destport = m.group(1)
    # (3) 送信
    self.send(str(msg), desthost, destport)

  def comp(self, requri, domain, ip, port):
    if requri.host == domain:
      return True
    if requri.host == ip:
      requri_port = requri.port
      if requri_port == None:
        requri_port = 5060
      elif requri_port == "":
        requri_port = 5060
      if str(requri_port) == str(port):
        return True
    return False

  def send(self, buf, host, port):
    if port == None:
      port = 5060
    elif port == "":
      port = 5060
    else:
      port = int(port)
    if self.ip == host and self.port == port:
      return
    self.sock.sendto(buf, 0, (host, port))
    print '>' * 72
    print 'send to ' + host + ' ' + str(port)
    print '-' * 8
    print buf

px = Proxy(sys.argv[1], sys.argv[2], int(sys.argv[3]))
print "ok"
px.start()    

File: xylitol.go
package main

import (
	"fmt"
	"log"
	"net"
	"os"
	"regexp"
	"strings"
)

var re = RegEx{}
var re_empty_line *regexp.Regexp
var re_crlf *regexp.Regexp
var re_wsp *regexp.Regexp
var re_start_line *regexp.Regexp
var re_hname *regexp.Regexp
var re_comma *regexp.Regexp

func InitRegEx() {
	re_empty_line = regexp.MustCompile(`(\r\n\r\n)|(\n\n)|(\r\r)`)
	re_crlf = regexp.MustCompile(`((\r\n)|(\n)|(\r))`)
	re_wsp = regexp.MustCompile(`((\r\n)|(\r)|(\n))[ \t]+`)
	re_start_line = regexp.MustCompile(`(([A-Z]+) ([^ ]+) )?SIP/2\.0( ([1-6][1-6][0-9]) ([^\r\n]+))?((\r\n)|(\r)|(\n))`)
	re_hname = regexp.MustCompile(`([^\s:]+)(\s*:\s*)`)
	re_comma = regexp.MustCompile(`\s*,\s*`)
}

type RegEx struct {
}

type Header struct {
	hname string
	hvals []string
}

func NewHeader() *Header {
	hdr := new(Header)
	hdr.hname = ""
	hdr.hvals = []string{}
	return hdr
}

type Message struct {
	method  string
	requri  string
	stcode  string
	reason  string
	headers []*Header
	body    string
}

func NewMessage() *Message {
	msg := new(Message)
	return msg
}

func (this *Message) Parse(s *string, addr *net.UDPAddr) bool {
	log.Println(*s)
	*s = strings.TrimLeft(*s, `\r\n`)
	// Empty-Line
	empty_line_pos := re_empty_line.FindStringIndex(*s)
	log.Println("Empty-Line Pos:", empty_line_pos[0])
	for i, v := range empty_line_pos {
		fmt.Println(i, v)
	}
	this.body = (*s)[empty_line_pos[1]:]
	fmt.Println("body=[" + this.body + "]")
	buf := re_wsp.ReplaceAllString((*s)[:empty_line_pos[0]], " ")
	buf = re_crlf.ReplaceAllString(buf, "\n")
	ary := strings.Split(buf, "\n")
	if this.ParseStartLine(&(ary[0])) == false {
		return false
	}
	for _, line := range ary[1:] {
		if this.ParseHeader(&line) == false {
			return false
		}
	}
	return true
}

func (this *Message)ParseStartLine(s *string) bool {
	sl := re_start_line.FindStringSubmatchIndex(*s)
	if len(sl) != 14 {
		fmt.Println(len(sl))
		return false
	}
	if sl[4] != -1 && sl[5] != -1 &&
		sl[6] != -1 && sl[7] != -1 {
		this.method = (*s)[sl[4]:sl[5]]
		this.requri = (*s)[sl[6]:sl[7]]
		this.stcode = ""
		this.reason = ""
	} else if sl[10] != -1 && sl[11] != -1 &&
		sl[12] != -1 && sl[13] != -1 {
		this.method = ""
		this.requri = ""
		this.stcode = (*s)[sl[10]:sl[11]]
		this.reason = (*s)[sl[12]:sl[13]]
	}
	fmt.Println("method=[" + this.method + "] requri=[" + this.requri + "]")
	fmt.Println("stcode=[" + this.stcode + "] reason=[" + this.reason + "]")
	return true
}

func (this *Message) ParseHeader(s *string) bool {
	pos := re_hname.FindStringSubmatchIndex(*s)
	hdr := NewHeader()
	hdr.hname = (*s)[:pos[3]]
	hdr.hvals = re_comma.Split((*s)[pos[1]:], -1)
	this.headers = append(this.headers, hdr)
	return true
}

func (this *Message)IsRequest() bool {
	if len(this.method) > 0 && len(this.requri) > 0 {
		return true
	}
	return false
}

func (this *Message)IsResponse() bool {
	if len(this.stcode) > 0 && len(this.reason) > 0 {
		return true
	}
	return false
}

func ProcRequest(msg *Message) {
}

func ProcResponse(msg *Message) {
}

func main() {
	fmt.Println("test")
	InitRegEx()
	addr := &net.UDPAddr{
		IP:   net.ParseIP("127.0.0.1"),
		Port: 5060,
	}
	sock, err := net.ListenUDP("udp", addr)
	if err != nil {
		log.Fatalln(err)
		os.Exit(1)
	}
	buf := make([]byte, 0xffff)
	for {
		len, addr, err := sock.ReadFromUDP(buf)
		if err == nil {
			s := string(buf[:len])
			log.Println(addr)
			msg := NewMessage()
			if msg.Parse(&s, addr) == false {
				continue
			}
			if msg.IsRequest() {
				ProcRequest(msg)
			} else if msg.IsResponse() {
				ProcResponse(msg)
			}
		}
	}
}
